src/reader.cpp

```diff
@@ -1743,8 +1743,17 @@ static void reader_interactive_init() {
         // Eventually we just give up and assume we're orphaend.
         for (unsigned long loop_count = 0;; loop_count++) {
             pid_t owner = tcgetpgrp(STDIN_FILENO);
-            shell_pgid = getpgrp();
+            // 0 is a valid return code from `tcsetpgrp()` under FreeBSD, and testing
```

Comment:
in which case... why doesn't mine return 0?

It says it "will return the foreground process group ID of that terminal if there is one, and some value larger than 1 that is not presently a process group ID otherwise". Which just strikes me as quite stupid, and just asking for TOCTTOUs. Why not just 0?

Reply:
Yup. Thanks.

Reply:
0 is actually a valid pid for the top-level process (`init` under Linux, `kernel` under FreeBSD).

Under Linux in my test cases I was getting -1, but when SSH'd to FreeBSD I was getting zero. The FreeBSD man page has this to say:

```
     The tcgetpgrp() function returns the value of the process group ID of the foreground process group
     associated with the terminal device.  If there is no foreground process group, tcgetpgrp() returns an
     invalid process ID.
```

A process gets assigned a parent id of zero when its own parent dies (giving parentship of the process to init, whose job is to then call `waitpid` to kill the zombie process). I'm not sure if that's what was happening in my test case (or how that could be what was happening in my case, and I'm suspecting that FreeBSD is returning zero to imply invalid.

I should clarify that comment.

Reply:
>0 is actually a valid pid for the top-level process (init under Linux, kernel under FreeBSD).

Init is always PID 1 in linux and [FreeBSD](https://www.freebsd.org/doc/handbook/basics-processes.html). PID 0 on linux is the swapper and scheduler, but those don't have children or terminals, so we shouldn't ever have to care about them. I'm not sure if they actually "do processy stuff" on BSD.

>A process gets assigned a parent id of zero when its own parent dies

On linux it gets 1, as above. And I'm pretty sure that applies to BSD too, since I'm pretty sure that also reparents to init.

Reply:
Sorry, I had the pid for init wrong. FreeBSD does reparent to 1 as well. Which makes this even more bizarre, and lends credence to the theory that 0 does indeed represent invalid under FreeBSD.. yet indicates that a subsequent attempt at `tcsetpgrp` will succeed.